libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,973 @@
+	}
+
+	if (!m_compilerStack)
+		compile(*file);
+	solAssert(m_compilerStack.get() != nullptr, "");
+
+	auto const sourceNode = findASTNode(dpos.position, dpos.path);
+	if (!sourceNode)
+	{
+		Json::Value emptyResponse = Json::arrayValue;
+		m_client->reply(_id, emptyResponse); // reply with "No references".
+		return;
+	}
+
+	auto locations = vector<SourceLocation>{};
+	if (auto const sourceIdentifier = dynamic_cast<Identifier const*>(sourceNode))
+	{
+		frontend::SourceUnit const& sourceUnit = m_compilerStack->ast(dpos.path);
+
+		if (auto decl = sourceIdentifier->annotation().referencedDeclaration)
+			findAllReferences(decl, decl->name(), sourceUnit, locations);
+
+		for (auto const decl: sourceIdentifier->annotation().candidateDeclarations)
+			findAllReferences(decl, decl->name(), sourceUnit, locations);
+	}
+	else if (auto const decl = dynamic_cast<VariableDeclaration const*>(sourceNode))
+	{
+		frontend::SourceUnit const& sourceUnit = m_compilerStack->ast(dpos.path);
+		findAllReferences(decl, decl->name(), sourceUnit, locations);
+	}
+	else if (auto const* functionDefinition = dynamic_cast<FunctionDefinition const*>(sourceNode))
+	{
+		frontend::SourceUnit const& sourceUnit = m_compilerStack->ast(dpos.path);
+		findAllReferences(functionDefinition, functionDefinition->name(), sourceUnit, locations);
+	}
+	else if (auto const* enumDef = dynamic_cast<EnumDefinition const*>(sourceNode))
+	{
+		frontend::SourceUnit const& sourceUnit = m_compilerStack->ast(dpos.path);
+		findAllReferences(enumDef, enumDef->name(), sourceUnit, locations);
+	}
+	else if (auto const memberAccess = dynamic_cast<MemberAccess const*>(sourceNode))
+	{
+		if (Declaration const* decl = memberAccess->annotation().referencedDeclaration)
+		{
+			frontend::SourceUnit const& sourceUnit = m_compilerStack->ast(dpos.path);
+			findAllReferences(decl, memberAccess->memberName(), sourceUnit, locations);
+		}
+	}
+	else
+		trace("references: not an identifier: "s + typeid(*sourceNode).name());
+
+	Json::Value jsonReply = Json::arrayValue;
+	for (SourceLocation const& location: locations)
+		jsonReply.append(toJson(location));
+
+	m_client->reply(_id, jsonReply);
+}
+
+void LanguageServer::log(string const& _message)
+{
+	if (m_trace < Trace::Messages)
+		return;
+
+	Json::Value json = Json::objectValue;
+	json["type"] = static_cast<int>(Trace::Messages);
+	json["message"] = _message;
+
+	m_client->notify("window/logMessage", json);
+
+	if (m_logger)
+		m_logger(_message);
+}
+
+void LanguageServer::trace(string const& _message)
+{
+	if (m_trace < Trace::Verbose)
+		return;
+
+	Json::Value json = Json::objectValue;
+	json["type"] = static_cast<int>(Trace::Verbose);
+	json["message"] = _message;
+
+	m_client->notify("window/logMessage", json);
+
+	if (m_logger)
+		m_logger(_message);
+}
+
+void LanguageServer::handleMessage(Json::Value const& _jsonMessage)
+{
+	string const methodName = _jsonMessage["method"].asString();
+
+	MessageId const id = _jsonMessage["id"].isInt()
```

Comment:
Why not use `std::optional<Json::Value>` for `MessageId`?

Reply:
I've just rewritten that part. `MessageId` now is just a `std::string`. When ID is passed as number, it's just converted to string. (I never experienced anything else than numberical IDs until now anyways). Also, the no-id case can now be represented as empty string.

Reply:
hm, didn't work. yes, I could do this what you proposed.